### PR TITLE
fix: documenting that kube-proxy only use the first address in each endpoint

### LINKED ISSUE #106267: kube-proxy only looks at Addresses[0] in EndpointSlice
### What happened?

An `EndpointSlice` can contain up to 1000 `Endpoints`, which can each contain up to 100 `Addresses`. Or at least, that's what the docs say, but actually, kube-proxy (or more specifically, `pkg/proxy/endpointslicecache.go`) only looks at the first Address in each Endpoint:

```
		endpointInfo := newBaseEndpointInfo(endpoint.Addresses[0], nodeName, zone, portNum, isLocal,
			endpoint.Ready, endpoint.Serving, endpoint.Terminating, endpoint.ZoneHints)
```

(This is compatible with how `EndpointSliceController` works, since it would only generate multi-Address Endpoints if a pod had multiple `PodIPs` of the same address family, which is not allowed. But it's not compatible with manually-generated EndpointSlices.)

### What did you expect to happen?

Either all `Addresses` are used, or else the docs state that only `Addresses[0]` is used and the field is an array solely for historical reasons.

### PR DESCRIPTION
**Pull request by cyclinder**
Signed-off-by: cyclinder <qifeng.guo@daocloud.io>

#### What type of PR is this?

/kind bug
#### What this PR does / why we need it:

the docs say:  "each endpoint in a `EndpointSlice`contains at least one but no more than 100 addresses", but currently, `kube-proxy` only use the first address in each endpoint and ignores additional addresses. so we should documenting it. 

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/106267

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by aojea** [comment]
/assign @robscott @thockin @danwinship 

kube-proxy is an implementation detail, should it be part of the APIs documentation?
I prefer not, but defer to the assignees

**Comment by thockin** [comment]
Discussion on #106267

**Comment by cyclinder** [comment]
>

**Comment by thockin** [comment]
My last comment in #106267 was: I think the simplest and most compat approach is to say "these are all assumed to be fungible and clients may choose to only use the first element".  We should not mention kube-proxy by name

**Comment by cyclinder** [comment]
hi @thockin, so we should only change API doc？

File: staging/src/k8s.io/api/discovery/v1/generated.proto [hunk 1]
```diff
@@ -34,8 +34,10 @@ message Endpoint {
   // addresses of this endpoint. The contents of this field are interpreted
   // according to the corresponding EndpointSlice addressType field. Consumers
   // must handle different types of addresses in the context of their own
-  // capabilities. This must contain at least one address but no more than
-  // 100.
+  // capabilities. This must contain at least one address but no more than 100.
+  // Warning: Currently. kube-proxy only use the first address in each endpoint
```
Comment (acumino) [inline staging/src/k8s.io/api/discovery/v1/generated.proto:38]:
```suggestion
  // Warning: Currently kube-proxy only use the first address in each endpoint
```

File: staging/src/k8s.io/api/discovery/v1/types.go [hunk 2]
```diff
@@ -71,8 +71,10 @@ type Endpoint struct {
 	// addresses of this endpoint. The contents of this field are interpreted
 	// according to the corresponding EndpointSlice addressType field. Consumers
 	// must handle different types of addresses in the context of their own
-	// capabilities. This must contain at least one address but no more than
-	// 100.
+	// capabilities. This must contain at least one address but no more than 100.
+	// Warning: Currently. kube-proxy only use the first address in each endpoint
```
Comment (acumino) [inline staging/src/k8s.io/api/discovery/v1/types.go:75]:
```suggestion
	// Warning: Currently kube-proxy only use the first address in each endpoint
```

File: staging/src/k8s.io/api/discovery/v1beta1/generated.proto [hunk 3]
```diff
@@ -34,8 +34,10 @@ message Endpoint {
   // addresses of this endpoint. The contents of this field are interpreted
   // according to the corresponding EndpointSlice addressType field. Consumers
   // must handle different types of addresses in the context of their own
-  // capabilities. This must contain at least one address but no more than
-  // 100.
+  // capabilities. This must contain at least one address but no more than 100.
+  // Warning: Currently. kube-proxy only use the first address in each endpoint
```
Comment (acumino) [inline staging/src/k8s.io/api/discovery/v1beta1/generated.proto:38]:
```suggestion
  // Warning: Currently kube-proxy only use the first address in each endpoint
```

File: staging/src/k8s.io/api/discovery/v1beta1/types.go [hunk 4]
```diff
@@ -75,8 +75,10 @@ type Endpoint struct {
 	// addresses of this endpoint. The contents of this field are interpreted
 	// according to the corresponding EndpointSlice addressType field. Consumers
 	// must handle different types of addresses in the context of their own
-	// capabilities. This must contain at least one address but no more than
-	// 100.
+	// capabilities. This must contain at least one address but no more than 100.
+	// Warning: Currently. kube-proxy only use the first address in each endpoint
```
Comment (acumino) [inline staging/src/k8s.io/api/discovery/v1beta1/types.go:79]:
```suggestion
	// Warning: Currently kube-proxy only use the first address in each endpoint
```

**Review by acumino — COMMENTED** [review]
Nit

**Comment by thockin** [comment]
@cyclinder ping me when this is updated

**Comment by cyclinder** [comment]
> @cyclinder ping me when this is updated

okay, I encountered some errors while generating the proto.pb file, I'm working on it

**Comment by cyclinder** [comment]
at first. I use `./hack/update-generated-protobuf.sh`,but failed:
```shell
#3 [internal] load metadata for k8s.gcr.io/build-image/kube-cross:v1.24.0-go1.17.5-bullseye.0
#3 ERROR: failed to do request: Head "https://k8s.gcr.io/v2/build-image/kube-cross/manifests/v1.24.0-go1.17.5-bullseye.0": Service Unavailable
------
 > [internal] load metadata for k8s.gcr.io/build-image/kube-cross:v1.24.0-go1.17.5-bullseye.0:
------
error: failed to solve: failed to solve with frontend dockerfile.v0: failed to create LLB definition: failed to do request: Head "https://k8s.gcr.io/v2/build-image/kube-cross/manifests/v1.24.0-go1.17.5-bullseye.0": Service Unavailable

To retry manually, run:

DOCKER_CLI_EXPERIMENTAL=enabled docker buildx build --load -t kube-build:build-307eb748e6-5-v1.24.0-go1.17.5-bullseye.0 --pull=false --build-arg=KUBE_CROSS_IMAGE=k8s.gcr.io/build-image/kube-cross --build-arg=KUBE_CROSS_VERSION=v1.24.0-go1.17.5-bullseye.0 /Users/cyclinder/Desktop/code/kubernetes/_output/images/kube-build:build-307eb748e6-5-v1.24.0-go1.17.5-bullseye.0
```
and then I try to  exec `./hack/update-generated-protobuf-dockerized.sh k8s.io/api/discovery/v1`, But it looks as if the generated proto.pb file is not as expected, is it a problem with the protoc version?
```shell
root@CycLinder-3 ~/Desktop/code/kubernetes$ protoc --version                                                                                                       ✹addresses_in_endpointslice 
libprotoc 3.19.1
```

File: staging/src/k8s.io/api/discovery/v1/generated.proto [hunk 5]
```diff
@@ -34,8 +34,9 @@ message Endpoint {
   // addresses of this endpoint. The contents of this field are interpreted
   // according to the corresponding EndpointSlice addressType field. Consumers
   // must handle different types of addresses in the context of their own
-  // capabilities. This must contain at least one address but no more than
-  // 100.
+  // capabilities. This must contain at least one address but no more than 100.
+  // these are all assumed to be fungible and clients may choose to only use the first element.
```
Comment (robscott) [inline staging/src/k8s.io/api/discovery/v1/generated.proto:38]: Nit: Wrap lines at 80 chars here and elsewhere.

File: staging/src/k8s.io/api/discovery/v1/types.go [hunk 6]
```diff
@@ -72,8 +72,9 @@ type Endpoint struct {
 	// addresses of this endpoint. The contents of this field are interpreted
 	// according to the corresponding EndpointSlice addressType field. Consumers
 	// must handle different types of addresses in the context of their own
-	// capabilities. This must contain at least one address but no more than
-	// 100.
+	// capabilities. This must contain at least one address but no more than 100.
+	// these are all assumed to be fungible and clients may choose to only use the first element.
+	// refer to: https://github.com/kubernetes/kubernetes/issues/106267
```
Comment (thockin) [inline staging/src/k8s.io/api/discovery/v1/types.go:77]: How about `https://issue.k8s.io/106267` as the URL?

File: staging/src/k8s.io/api/discovery/v1/generated.proto (hunk 5 above)
Comment (cyclinder) [reply staging/src/k8s.io/api/discovery/v1/generated.proto]: thanks, Done

File: staging/src/k8s.io/api/discovery/v1/types.go (hunk 6 above)
Comment (cyclinder) [reply staging/src/k8s.io/api/discovery/v1/types.go]: it's better :)

**Comment by cyclinder** [comment]
Thanks for your review!  can you take a look?
/cc @thockin @robscott

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Pull request merged**